text/0000-unified_coroutines.md

```diff
@@ -0,0 +1,411 @@
+}
+```
+If we introduce this trait, and rename the original `Generator::resume` method to `Generator::resume_with_args`, the existing behavior will not change. But we think this approach is **WRONG**, and the Generator trait should stay with the changes proposed above (`resume`/`call_resume` accepting a tuple). The rationale for this decision is provided in [future-possibilities] section.
+
+3. Need to teach the special interaction between generator arguments and the yield statement.
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+Rationale:
+- This introduces the missing piece into the generators, rounding out an unfinished feature.
+- Unifies 2 parts of the language, improving laungage coherence
+- Allows implementations of new patterns, such as complex state machines, which are an extremely useful tool in several areas, eg. network protocol implementations.
+
+Alternatives:
+- Only current alternative is storing required information inside side channels such as `Rc<RefCell<Args>>` 
+or a thread local storage, which introduces runtime overhead and requires `std`. Proposed design is `no_std` compatible.
+- The proposed syntax could be changed, but from the explored options, we pose that the simplest syntax is best, even though it introduces new semantics.
+- Leave the generator as is, leaving it disconnected from the rest of the language.
+- Remove generators completely 
+
+# Prior art
+[prior-art]: #prior-art
+
+- Pre-RFC published by different author on rust-internals forum [link](https://internals.rust-lang.org/t/pre-rfc-generator-resume-args/10011/5)
+   
+  Explored the design space and proposed a basic design of the modified `Generator trait`. The described approach solved only the 'Generator resume arguments' part of this RFC, and did not attempt to unify generators with closures, which resulted in unnecessarily complex design, which kept the generators as a separate concept, and even introduced another layer of complexity in form of a trait alias for generators with/without arguments. But nonetheless, reading this discussion was an invaluable source of ideas in this design space.
+  
+- Work on implementing futures which wouldn't require TLS: [link](https://github.com/rust-lang/rust/issues/62918)
+
+- Python & Lua coroutines - They can be resumed with arguments, with yield expression returning these values [usage](https://www.tutorialspoint.com/lua/lua_coroutines.htm). 
+  
+  These are interesting, since they both adopt a syntax, in which the yield expression returns values passed to resume. We think that this approach is right one for dynamic languages like Python or lua but the wrong one for Rust. The reason is, these languages are dynamically typed, and allow passing of multiple values into the coroutine. The design proposed here is static, and allows passing only a single argument into the coroutine, a tuple. The argument tuple is treated the same way as in the `Fn*` family of traits. 
```

Comment:
I don't believe dynamic typing really changes things here. It's true that in dynamic languages you *can* come up with elaborate protocols where each resume takes a different set of argument types, but that's rare and confusing- most use cases stick with a single type, just as we are enforcing in Rust generators.

Letting `yield` evaluate to the passed-in values, with the expression's type determined by the closure-syntax argument types, shouldn't cause any problems. Indeed, it simplifies things by introducing fewer new semantics (see above and below).

Reply:
Well, it does not introduce new cognitive load but introduces fallible syntax, and makes the default/shorter code the wrong one in many cases.

Reply:
We already have falliable syntax. 

```rust
let x = {
    let mut i = 2;
    loop {
        i += if i > 16 {
            i/2
        } else {
            break i
        };
    }
};
```